### ISSUE TITLE
fireEvent fails to trigger events when combining render and renderHook in a single test

### ISSUE THREAD

**Issue by WookieFPV** [callstack/react-native-testing-library#1703]
## Describe the bug

When using both `render(Component)` and `renderHook` in a test, they do not work well together:

- If `renderHook` is used before `render`, the test works as expected in my case.
- If `render` is called first, followed by `renderHook`, then `fireEvent` does not trigger events as expected (it becomes a no-op).

This issue began in version 18.2.1 (it worked correctly in 18.2.0). The following change appears related: [GitHub Link](https://github.com/callstack/react-native-testing-library/pull/1696/files#diff-4e4de45566536ffba3b6658471591d81d3a40cfd00b08b4849db707565efbabd).

**Root Cause Analysis**

Inside `fireEvent`, there is a check to see if the element is mounted (`isElementMounted()`). This check currently fails, causing `fireEvent` to return early. Disabling this check allows my test to pass, but it may reveal deeper underlying issues.

**Questions**

- Is it supported or recommended to use both `render` and `renderHook` in the same test?

## Expected behavior

fireEvent triggers the event

## Steps to Reproduce

```typescript
import { fireEvent, render, renderHook } from '@testing-library/react-native';
import { Pressable, Text } from 'react-native';
import { create } from 'zustand';

type ToggleStore = {
    active: boolean;
    setActive: (active: boolean) => void;
};

const useToggleStore = create<ToggleStore>((set) => ({
    active: false,
    setActive: (bool) => set(() => ({ active: bool })),
}));

const ExampleToggle = () => {
    const { active, setActive } = useToggleStore();
    return (
        <Pressable testID={'btn'} onPress={() => setActive(!active)}>
            <Text>{`State: ${active}`}</Text>
        </Pressable>
    );
};

describe('Tests with render(Component) & renderHook inside a single test', () => {
    // The only difference in this two tests is the order of render & renderHook
    // I am not sure if using both render & renderHook inside a single tests is a good idea tbh

    // This tests runs fine pre 18.2.1
    // see related change: https://github.com/callstack/react-native-testing-library/pull/1696/files#diff-4e4de45566536ffba3b6658471591d81d3a40cfd00b08b4849db707565efbabd
    test('render before renderHook => ERROR', () => {
        const view = render(<ExampleToggle />);
        const { result } = renderHook(() => useToggleStore());

        expect(view.getByText(`State: ${false}`)).toBeVisible();
        expect(result.current.active).toEqual(false);

        fireEvent.press(view.getByTestId('btn')); // this is a noop because of the isElementMounted() check inside of fireEvent returns false which aborts the function

        expect(view.getByText(`State: ${true}`)).toBeVisible(); // Error: Unable to find an element with text: State: true
        expect(result.current.active).toEqual(true);
    });

    test('renderHook before render => WORKS', () => {
        const { result } = renderHook(() => useToggleStore());
        const view = render(<ExampleToggle />);

        expect(view.getByText(`State: ${false}`)).toBeVisible();
        expect(result.current.active).toEqual(false);

        fireEvent.press(view.getByTestId('btn'));

        expect(view.getByText(`State: ${true}`)).toBeVisible();
        expect(result.current.active).toEqual(true);
    });
});

```

## Versions

@testing-library/react-native: 12.8.1 => 12.8.1
react: 18.3.1 => 18.3.1
react-native: 0.75.4 => 0.75.4
react-test-renderer: 18.3.1 => 18.3.1

**Comment by mdjastrzebski** [comment]
TLDR: you should not mix two renders ('render' or `renderHook` calls in one test).

Longer explanation:
- when you can `render` or `renderHook` it attaches the render result to the `screen` object, so you can use it to query active tree.
- next render call, removes the previous render output from the screen object and attaches the new one. The old rendered element tree is no longer reachable through the screen object.
- some matchers inside RNTL, e.g. `toBeOnTheScreen` rely on the `screen` object to reference the currently rendered tree. If you call `render` (or `renderHook`) in the meantime, the original element tree is no longer considered to be mounted, and such marchers will fail.
